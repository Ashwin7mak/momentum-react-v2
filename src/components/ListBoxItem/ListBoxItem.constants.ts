const CLASS_PREFIX = 'md-list-box-item';

const STYLE = {
  tickIcon: `${CLASS_PREFIX}-tick-icon`,
};

export { STYLE };
