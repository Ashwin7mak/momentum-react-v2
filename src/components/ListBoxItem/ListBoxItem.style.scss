.md-list-box-item-wrapper {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5rem;
  cursor: pointer;
  color: var(--theme-text-primary-normal);
  border-radius: 0.5rem;
  transition: background-color 0.2s, color 0.2s, border-color 0.2s;

  &:focus {
    box-shadow: var(--focus-ring-box-shadow);
    outline: var(--focus-ring-outline);
  }

  &:hover {
    background-color: var(--theme-background-primary-hover);
  }

  &:active {
    background-color: var(--theme-background-primary-active);
  }

  .md-list-box-item-tick-icon {
    fill: var(--theme-control-active-normal);
  }
}

.md-list-box-item-contents-wrapper {
  display: flex;
  justify-content: flex-start;
  align-items: center;

  > *:not(:first-child) {
    margin-left: 0.75em;
  }
}