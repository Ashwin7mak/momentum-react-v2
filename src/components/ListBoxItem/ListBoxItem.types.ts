import { Node } from '@react-types/shared';
import { CSSProperties } from 'react';

export interface Props<T> {
  /**
   * Represents the item to be displayed in this ListBoxItem
   */
  item: Node<T>;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;
}
