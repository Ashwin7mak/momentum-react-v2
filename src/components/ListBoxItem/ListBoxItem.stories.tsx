import React, { FC } from 'react';
import { Story } from '@storybook/react';

import ListBoxItem, { ListBoxItemProps } from './';
import {
  Title,
  Subtitle,
  Description,
  Primary,
  ArgsTable,
  PRIMARY_STORY,
} from '@storybook/addon-docs';

import Documentation from './ListBoxItem.documentation.mdx';

const DocsPage: FC = () => (
  <>
    <Title />
    <Subtitle />
    <Description />
    <Documentation />
    <Primary />
    <ArgsTable story={PRIMARY_STORY} />
  </>
);

export default {
  title: 'Momentum UI/ListBoxItem',
  component: ListBoxItem,
  parameters: {
    expanded: true,
    docs: {
      page: DocsPage,
    },
  },
  argTypes: {
    propName: {
      defaultValue: undefined,
      description: 'Description goes here.',
      options: [undefined, 'Option 1', 'Option 2'],
      control: { type: 'select' },
      table: {
        type: {
          summary: 'string',
        },
        defaultValue: {
          summary: undefined,
        },
      },
    },
  },
};

const MultiTemplate: Story<ListBoxItemProps> = (args: ListBoxItemProps, { parameters }) => {
  const { variants } = parameters;

  const items = variants.map((variant, index: number) => (
    <div key={index}>
      <ListBoxItem {...args} {...variant} />
      <p>{variant.label}</p>
    </div>
  ));

  return (
    <div
      style={{
        display: 'grid',
        gridTemplateColumns: `repeat(4, auto)`,
        gap: '1.5rem',
        alignItems: 'end',
      }}
    >
      {items}
    </div>
  );
};

const Template: Story<ListBoxItemProps> = (args) => <ListBoxItem {...args} />;

const Example = Template.bind({});

Example.args = {
  propName: 'Value 1',
};

const Common = MultiTemplate.bind({});

Common.parameters = {
  variants: [
    {},
    { propName: 'Value 1', label: 'With value 1' },
    { propName: 'Value 2', label: 'With value 2' },
  ],
};

export { Example, Common };
