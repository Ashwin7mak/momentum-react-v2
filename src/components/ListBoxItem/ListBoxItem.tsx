/* eslint-disable @typescript-eslint/ban-types */
import React, { ReactElement, useContext } from 'react';

import './ListBoxItem.style.scss';
import { Props } from './ListBoxItem.types';
import { STYLE } from './ListBoxItem.constants';
import { ListBoxContext } from '../ListBoxBase/ListBoxBase';
import { useOption } from '@react-aria/listbox';
import Icon from '../Icon';
import ListItemBaseSection from '../ListItemBaseSection';
import ListItemBase from '../ListItemBase';

function ListBoxItem<T>(props: Props<T>): ReactElement {
  const { item, style } = props;
  const ref = React.useRef<HTMLLIElement>(null);
  const { state, shouldVirtualizeItems, shouldWrapItems, shouldItemFocusBeInset } =
    useContext(ListBoxContext);

  const isDisabled = state.disabledKeys.has(item.key);
  const isSelected = state.selectionManager.isSelected(item.key);

  const { optionProps } = useOption(
    {
      key: item.key,
      isDisabled,
      'aria-label': item['aria-label'],
      isSelected,
      shouldSelectOnPressUp: true,
      shouldFocusOnHover: false,
      isVirtualized: shouldVirtualizeItems,
    },
    state,
    ref
  );

  if (shouldWrapItems) {
    return (
      <ListItemBase
        style={style}
        key={item.key}
        ref={ref}
        {...optionProps}
        isDisabled={isDisabled}
        shouldItemFocusBeInset={shouldItemFocusBeInset}
      >
        <ListItemBaseSection position="fill">{item.rendered}</ListItemBaseSection>
        {isSelected && (
          <ListItemBaseSection position="end">
            <Icon className={STYLE.tickIcon} name="check" weight="bold" scale={16} />
          </ListItemBaseSection>
        )}
      </ListItemBase>
    );
  } else {
    if (React.isValidElement(item.rendered)) {
      return React.cloneElement(item.rendered as ReactElement, {
        key: item.key,
        ref,
        isDisabled,
        isSelected,
        style,
        ...optionProps,
      });
    } else {
      return <>{item.rendered}</>;
    }
  }
}

/**
 * ListBoxItem component used internally as a wrapper for items inside a listbox.
 * It's using the ListItem component internally.
 * @internal
 */

export default ListBoxItem;
