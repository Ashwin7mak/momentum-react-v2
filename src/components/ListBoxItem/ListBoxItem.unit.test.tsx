import ListBoxItem from '.';
import { mount } from 'enzyme';
import React from 'react';
import { useListState } from '@react-stately/list';
import { renderHook } from '@testing-library/react-hooks';
import { Item } from '@react-stately/collections';

jest.mock('@react-aria/listbox', () => {
  return {
    useOption: () => ({ optionProps: {} }),
  };
});

describe('ListBoxItem', () => {
  const { result } = renderHook(() =>
    useListState({
      children: [
        <Item key="$.0" aria-label="0">
          Item 1
        </Item>,
        <Item key="$.1" aria-label="1">
          Item 2
        </Item>,
      ],
    })
  );

  const state = result.current;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let useContextMock: jest.Mock<any, any>;

  beforeEach(() => {
    useContextMock = React.useContext = jest.fn();
  });

  describe('snapshot', () => {
    it('should match snapshot', () => {
      useContextMock.mockReturnValue({ state, shouldWrapItems: true });
      const item = state.collection.getItem('$.0');

      const wrapper = mount(<ListBoxItem key={item.key} item={item} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should render the given item', () => {
      useContextMock.mockReturnValue({ state, shouldWrapItems: true });
      const item = state.collection.getItem('$.1');

      const wrapper = mount(<ListBoxItem key={item.key} item={item} />);

      const element = wrapper.find('li div').getDOMNode();

      expect(element.innerHTML).toBe(item.rendered);
    });
  });
});
