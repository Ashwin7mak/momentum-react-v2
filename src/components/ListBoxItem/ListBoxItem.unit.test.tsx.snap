// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBoxItem snapshot should match snapshot 1`] = `
<ListBoxItem
  item={
    Object {
      "aria-label": "0",
      "childNodes": Object {
        Symbol(Symbol.iterator): [Function],
      },
      "hasChildNodes": false,
      "index": 0,
      "key": "$.0",
      "level": 0,
      "nextKey": "$.1",
      "parentKey": null,
      "prevKey": undefined,
      "props": Object {
        "aria-label": "0",
        "children": "Item 1",
      },
      "rendered": "Item 1",
      "shouldInvalidate": undefined,
      "textValue": "Item 1",
      "type": "item",
      "value": undefined,
      "wrapper": undefined,
    }
  }
>
  <ListItemBase
    isDisabled={false}
    key="$.0"
  >
    <FocusRing>
      <FocusRing
        focusClass="md-focus-ring-wrapper children"
        focusRingClass="md-focus-ring-wrapper children"
      >
        <li
          className="md-list-item-base-wrapper"
          data-disabled={false}
          data-shape="rectangle"
          data-size={40}
          onBlur={[Function]}
          onFocus={[Function]}
          role="listitem"
        >
          <ListItemBaseSection
            position="fill"
          >
            <div
              data-position="fill"
            >
              Item 1
            </div>
          </ListItemBaseSection>
        </li>
      </FocusRing>
    </FocusRing>
  </ListItemBase>
</ListBoxItem>
`;
